workflow/rules/utils.smk

```diff
@@ -33,4 +33,29 @@ rule tabix_known_variants:
         get_tabix_params
     cache: True
     wrapper:
-        "0.45.1/bio/tabix"
\ No newline at end of file
+        "0.45.1/bio/tabix"
+
+rule bam_regions:
+    input:
+        ref="resources/genome.fasta",
+        bam="results/recal/{sample}.sorted.bam"
+    output:
+        temp("results/regions/{sample}.bed")
+    log:
+        "logs/bam-regions/{sample}.log"
+    conda:
+        "../envs/bedregions.yaml"
+    shell:
+        "samtools view -b {input.bam} | genomeCoverageBed -ibam stdin -g {input.ref} > {output} 2> {log}"
```

Comment:
Just to confirm, this will output a bed file with only those regions with at least a coverage of 1?

Reply:
Indeed it won't.
So [bamtobed](https://bedtools.readthedocs.io/en/latest/content/tools/bamtobed.html) might be what we are looking for.

I just tested bamtobed for 10 Reads. The output looks like this:
```
1	21707	21850	M05890:70:000000000-CGM9D:1:1114:20746:10100/1	0	+
1	21707	21850	M05890:70:000000000-CGM9D:1:1114:20746:10100/2	0	-
1	31842	31984	M05890:70:000000000-CGM9D:1:2111:25061:8808/1	0	+
1	31842	31986	M05890:70:000000000-CGM9D:1:2111:25061:8808/2	0	-
1	189965	190072	M05890:70:000000000-CGM9D:1:1105:9369:15622/2	0	+
1	189965	190111	M05890:70:000000000-CGM9D:1:1107:17171:18596/1	23	+
1	189965	190111	M05890:70:000000000-CGM9D:1:1109:2264:16933/2	23	+
1	189965	190072	M05890:70:000000000-CGM9D:1:1105:9369:15622/1	0	-
1	190003	190143	M05890:70:000000000-CGM9D:1:1107:17171:18596/2	23	-
1	190003	190143	M05890:70:000000000-CGM9D:1:1109:2264:16933/1	23	-
```

I guess the first three columns are what we would expect.

Reply:
`samtools view -b <bam> | bedtools bamtobed -i stdin | bedtools merge -i stdin` would end up like this for the example output above: 
```
1	21707	21850
1	31842	31986
1	189965	190143
```

Reply:
But bamtobed will generate a huge overhead with one entry per read, as pure text. I think mosdepth is the way to go. It can output regions, that are even quantized, such that we can control the resolution. Probably we can even quantize into `0` and `>0`. https://github.com/brentp/mosdepth

Reply:
Wait, maybe bedtools genomecov works as well. https://bedtools.readthedocs.io/en/latest/content/tools/genomecov.html. You can use -max 1, to bin all regions with at least a coverage of 1 into a single bin.

Reply:
That would be preferable to mosdepth, cause it is easier to pipe.

Reply:
Using the `-max 1` creates some weird output:

```1	0	248955957	248956422	0.999998
1	1	465	248956422	1.8678e-06
10	0	133797422	133797422	1
11	0	135086622	135086622	1
12	0	133275309	133275309	1
13	0	114364328	114364328	1
14	0	107043718	107043718	1
15	0	101991189	101991189	1
16	0	90338345	90338345	1
17	0	83257441	83257441	1
18	0	80373285	80373285	1
19	0	58617616	58617616	1
2	0	242193529	242193529	1
20	0	64444167	64444167	1
21	0	46709983	46709983	1
22	0	50818468	50818468	1
3	0	198295559	198295559	1
4	0	190214555	190214555	1
5	0	181538259	181538259	1
6	0	170805979	170805979	1
7	0	159345973	159345973	1
8	0	145138636	145138636	1
9	0	138394717	138394717	1
MT	0	16569	16569	1
X	0	156040895	156040895	1
Y	0	57227415	57227415	1
...
```

Reply:
mosdepth requires a bed-file (specified by the `-by` parameter) as input to create the regions?

Reply:
What genomecov does, is give you the number of sites with 0 coverage per chromosome (0 in 2nd column) and as you requested `-max 1`, also the number of sites with coverage 1 or above. Check out the docs for a detailed description of the columns:
https://bedtools.readthedocs.io/en/latest/content/tools/genomecov.html

It seems like your sample only has coverage for very few sites on chromosome 1, which seems weird.

Reply:
If I understand correctly that you want a bed file with all the regions that have read coverage, I think what you want is [`bedtools merge`](https://bedtools.readthedocs.io/en/latest/content/tools/merge.html) that you can also do on BAM files to "merge" all reads into contiguous regions.

Reply:
So basically it should just work on the BAM file without any piping necessary:

```
bedtools merge -i {input.bam} > {output.bed}
```

Reply:
Ah yeah, mosdepth quantization also works! :D